core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -432,6 +435,57 @@ private[spark] class Executor(
   }
 
   /**
+   * Supervises the killing / cancellation of a task by sending the interrupted flag, optionally
+   * sending a Thread.interrupt(), and monitoring the task until it finishes.
+   */
+  private class TaskReaper(taskRunner: TaskRunner, interruptThread: Boolean) extends Runnable {
+
+    private[this] val killPollingFrequencyMs: Long =
+      conf.getTimeAsMs("spark.task.killPollingFrequency", "10s")
+
+    private[this] val killTimeoutMs: Long = conf.getTimeAsMs("spark.task.killTimeout", "2m")
+
+    private[this] val takeThreadDump: Boolean =
+      conf.getBoolean("spark.task.threadDumpKilledTasks", true)
+
+    override def run(): Unit = {
+      val startTimeMs = System.currentTimeMillis()
+      def elapsedTimeMs = System.currentTimeMillis() - startTimeMs
+
+      while (!taskRunner.isFinished && elapsedTimeMs < killTimeoutMs) {
+        taskRunner.kill(interruptThread = interruptThread)
+        taskRunner.synchronized {
+          Thread.sleep(killPollingFrequencyMs)
```

Comment:
Why is the sleep inside synchronized ?
Also, would be good to use wait/notify instead of sleep - so that the TaskReaper's are released proactively as soon as task finishes, so that number of threads in the cached pool does not go very high.